### PR TITLE
8262341: Refine identical code in AddI/LNode.

### PR DESCRIPTION
**Pull request by rwestrel**
AddINode::Ideal() and AddlNode::Ideal() are almost identical but the
same logic had to be duplicated because AddINode::Ideal() tests its
inputs for Op_AddI, Op_SubI etc. while AddLNode::Ideal() tests for
Op_AddL, Op_SubL etc. This patch refactors the code so the common
logic is in a single method parameterized by a BasicType argument.

The way I've done this before in the context of int/long counted loops
was to use and extra virtual method operates_on(). So:

n->Opcode() == Op_AddI becomes n->is_Add() && n->operates_on(T_INT)

Working on this change made me realize that pattern doesn't work that well:

- it's quite a bit more verbose and converting existing code is not as
  mechanical as we would like to avoid conversion errors.

- it breaks when a class has a subclass. For instance AddNode has
  OrINode and OrLNode as subclasses so testing for n->is_Add() returns
  true with an OrI node.

Instead, this change introduces new functions. For instance of
AddI/AddL:

int Op_Add(BasicType bt)

that returns either Op_AddI or Op_AddL depending on bt. This made
refactoring the AddINode::Ideal() logic straightforward. I removed all
use of operates_on() as well and converted existing code to the new
Op_XXX() functions.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262341](https://bugs.openjdk.java.net/browse/JDK-8262341): Refine identical code in AddI/LNode.

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 5160efe299990fc7a84b5335b39de13a4c501d83
 * [Eric Liu](https://openjdk.java.net/census#eliu) (@theRealELiu - Author) ⚠️ Review applies to 22292277edec10621d16ae67eb1950fd936f0878

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6607/head:pull/6607` \
`$ git checkout pull/6607`

Update a local copy of the PR: \
`$ git checkout pull/6607` \
`$ git pull https://git.openjdk.java.net/jdk pull/6607/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6607`

View PR using the GUI difftool: \
`$ git pr show -t 6607`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6607.diff">https://git.openjdk.java.net/jdk/pull/6607.diff</a>

</details>

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
Very nice, looks good to me.

File: src/hotspot/share/opto/addnode.cpp [hunk 1]
```diff
@@ -263,66 +263,69 @@ AddNode* AddNode::make(Node* in1, Node* in2, BasicType bt) {
...
     op1 = op2;
     op2 = in2->Opcode();
   }
-  if( op1 == Op_SubI ) {
-    const Type *t_sub1 = phase->type( in1->in(1) );
-    const Type *t_2    = phase->type( in2        );
-    if( t_sub1->singleton() && t_2->singleton() && t_sub1 != Type::TOP && t_2 != Type::TOP )
-      return new SubINode(phase->makecon( add_ring( t_sub1, t_2 ) ), in1->in(2) );
+  if (op1 == Op_Sub(bt)) {
+    const Type *t_sub1 = phase->type(in1->in(1));
```
Comment (theRealELiu) [inline src/hotspot/share/opto/addnode.cpp:280]: I'm not very clear about the current code style which we should follow. Shall we need to align the style in the changed code?

Comment (rwestrel) [reply src/hotspot/share/opto/addnode.cpp:280]: Are you commenting about the change from:
if( op1 to if (op1
?
The first style is used in a some places and the guideline is to switch to the second one.

Comment (theRealELiu) [reply src/hotspot/share/opto/addnode.cpp:280]: How about the style like `const Type *t_sub1`? Whether it should be `const Type* t_sub1`.

Comment (rwestrel) [reply src/hotspot/share/opto/addnode.cpp:280]: It should be const Type* t_sub1. I will fix it. Thanks.

**Review by theRealELiu — APPROVED** [review]
LGTM.

File: src/hotspot/share/opto/addnode.hpp [hunk 2]
```diff
@@ -51,7 +51,8 @@ class AddNode : public Node {
 
   // We also canonicalize the Node, moving constants to the right input,
   // and flatten expressions (so that 1+x+2 becomes x+3).
-  virtual Node *Ideal(PhaseGVN *phase, bool can_reshape);
+  virtual Node* Ideal(PhaseGVN *phase, bool can_reshape);
```
Comment (fg1417) [inline src/hotspot/share/opto/addnode.hpp:54]: Should be `PhaseGVN* phase` here?

Comment (rwestrel) [reply src/hotspot/share/opto/addnode.hpp]: Thanks for looking at this. You're right. Fixed.

**Pull request closed without merging**